### DISCUSSION TITLE
Could we support transaction when inserting or updating data?

### DISCUSSION THREAD

**Discussion by xuefeimiaoao** [milvus-io/milvus#21221] [Q&A and General discussion] [answered]
We'd like to updating collection with transaction operation, cus we will generate dirty data in case that our client is oom or else.

**Comment by yhmo** [comment, marked as answer]
No, we didn't implement transaction machinery in milvus. We ensure each insert or delete operation is atomic.

**Reply by xuefeimiaoao** [reply]
So do you mean that the single operation like `DB::InsertVectors(...)` will be succeeded atomically? And do we have any plan to support the transactional atomicity of one operation involving both upsertion and deletion operations furthermore?

**Reply by yhmo** [reply]
Yes, a single operation(insert/delete) is atomic. If the operation succeeds, the data persisted. Otherwise, nothing changes to the database.
Currently, we have no upsert interface(atomic operation). It is in developing. Now, users have to do two-step upsert: delete+insert, but this is not atomic.

**Reply by yhmo** [reply]
We don't intend to implement heavy transaction machinery to Milvus, because it is more like an OLAP database.

**Reply by xuefeimiaoao** [reply]
Thank you so much!
